File: frontend/src/types/task.ts
export type TaskStatus = "pending" | "in_progress" | "completed"
export type Priority = "low" | "medium" | "high" | "urgent"
export type RecurrenceType = "none" | "daily" | "weekly" | "monthly" | "custom"

export interface Task {
  id: string
  title: string
  description: string | null
  due_date: string | null
  priority: Priority
  status: TaskStatus
  category_id: string | null
  household_id: string | null
  assigned_to_id: string | null
  recurrence: RecurrenceType
  recurrence_rule: string | null
  tags: string[]
  created_at: string
  updated_at: string
}

export interface TaskFormValues {
  title: string
  description: string
  due_date: string
  priority: Priority
  category_id: string
  recurrence: RecurrenceType
  tags: string
}

export interface TaskPayload {
  title?: string
  description?: string | null
  due_date?: string | null
  priority?: Priority
  status?: TaskStatus
  category_id?: string | null
  household_id?: string | null
  assigned_to_id?: string | null
  recurrence?: RecurrenceType
  tags?: string[]
}

File: backend/app/models/task.py
import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, Enum, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base_class import Base
from app.models.enums import Priority, RecurrenceType, TaskStatus
from app.models.mixins import TimestampMixin, UUIDMixin


class Task(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "tasks"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    household_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id", ondelete="SET NULL"), nullable=True, index=True
    )
    assigned_to_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )
    category_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("categories.id", ondelete="SET NULL"), nullable=True, index=True
    )

    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text)
    due_date: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), index=True)
    priority: Mapped[Priority] = mapped_column(Enum(Priority, name="task_priority"), default=Priority.MEDIUM, nullable=False)
    status: Mapped[TaskStatus] = mapped_column(
        Enum(TaskStatus, name="task_status"), default=TaskStatus.PENDING, nullable=False, index=True
    )
    recurrence: Mapped[RecurrenceType] = mapped_column(
        Enum(RecurrenceType, name="task_recurrence"), default=RecurrenceType.NONE, nullable=False
    )
    recurrence_rule: Mapped[Optional[str]] = mapped_column(String(255))
    tags: Mapped[Optional[list[str]]] = mapped_column(ARRAY(String), default=list)
